### PR TITLE
More thorough build checks for Centipede

### PR DESCRIPTION
**Pull request by Alan32Liu**
Solves the CI failure in #8690, and uses the chance to add more thorough tests for the sanitized target binary and the unsanitized target binary for `Centipede`, which is something we thought about but did not have the chance to implement:
1. When building sanitized binaries with `helper.py` (i.e., local or GitHub CI): Unsanitized ones will be built automatically into the same docker container. Now bad_build_check tests both
    * a) if main fuzz targets can run with the auxiliary sanitized binaries, and 
    * b) if the auxiliaries are built with the correct sanitizers.
3. When In the Trial build and production build: Two kinds of binaries will be in separated buckets / docker containers. Now Script bad_build_check tests either
    * a) if the unsanitized binaries can run without the sanitized ones, or
    * b) if the sanitized binaries are built with the correct sanitizers.

### CODE REVIEW

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [inline infra/base-images/base-runner/test_all.py:174]: Can you put this `os.environ['SANITIZER'] = 'none'...os.environ['SANITIZER'] = sanitizer` in a try...finally so we can ensure it always gets reset properly.

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [inline infra/base-images/base-runner/test_all.py:165]: I'm confused why we are testing sanitized and unsanitized builds in the same build. Aren't these done completely separately?

Comment (jonathanmetzman) [reply infra/base-images/base-runner/test_all.py:165]: OK. I think I understand (though this is complicated). I think in helper.py you are doing both builds so that people can test locally. But I think in production you are keeping the builds seperate.
If i may suggest, maybe put the unsanitized build in a subdirectory, it will make it easier for people to repro crashes if their sanitized target is always in the same spot.

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
Strange: I cannot reproduce [edad2e6](https://github.com/google/oss-fuzz/pull/8697/commits/edad2e6ce96aa411af2ce575c552aaf752fd89a5)'s CI error.
I am not sure if I can fully understand it either: It complains that it cannot find the `centipede` binary. The only reason I can think of is somehow it did not properly reset env var `SANITIZER` and expects to find `centipede` in the sanitized binary's child directory to launch fuzzing.
Hence I added the `try`...`finally` block as suggested.

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
Reorganized the code to avoid setting ENV VAR, in case that somehow caused the CI failure that is unproducible locally.

**Comment by Alan32Liu** [comment]
Ah, I see now. It is still because CI builds do not build unsanitized binaries when building sanitized ones, and we don't copy binary `centipede` to sanitized build directories.
I will revert some of the commits tmr.

**Comment by jonathanmetzman** [comment]
> Ah, I see now. It is still because CI builds do not build unsanitized binaries when building sanitized ones,

It's important to be specific. The trial builds and production builds do not do this double build, the github CI and helper.py do.

**Comment by Alan32Liu** [comment]
> > Ah, I see now. It is still because CI builds do not build unsanitized binaries when building sanitized ones,
> 
> It's important to be specific. The trial builds and production builds do not do this double build, the github CI and helper.py do.

Yep, I should document this somewhere.
Would you happen to know any good place?

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer none --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
OK!
I reckon that shows my understanding is correct.
Let me make use of the trick in #8698.

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer none --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer none --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
@jonathanmetzman A gentle ping on this, as we need it to fix #8690 : )
Or would you prefer me to merge it into #8690 to test it out?
Thanks!

**Comment by jonathanmetzman** [comment]
> @jonathanmetzman A gentle ping on this, as we need it to fix #8690 : ) Or would you prefer me to merge it into #8690 to test it out? Thanks!

I was offline on monday and tuesday

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [inline infra/base-images/base-runner/test_all.py:223]: I think American spelling might be more common. "organize"

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [inline infra/base-images/base-runner/test_all.py:213]: I really don't like this. Pleas emake this function return an empty list or None when it doesn't need to return auxiliary fuzzers and don't bother passing the num_expected_targets so that the list is proper size.

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [inline infra/base-images/base-runner/test_all.py:190]: Start these with a period.

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [inline infra/base-images/base-runner/test_all.py:223]: the argument "directory" is not very descriptive.

Comment (jonathanmetzman) [inline infra/base-images/base-runner/test_all.py:223]: I think the reason this function name is vague is because the purpose of the function is vague and not so clear. I think it would be better if do_bad_build_check found the auxilliary target itself.

**Review by jonathanmetzman — CHANGES_REQUESTED** [review]
Just make do_bad_build_check find the auxiliary binary.

**Comment by jonathanmetzman** [comment]
> > @jonathanmetzman A gentle ping on this, as we need it to fix #8690 : ) Or would you prefer me to merge it into #8690 to test it out? Thanks!
> 
> I was offline on monday and tuesday. If there's an issue that doesn't effect production but is blocking you, feel free to merge it rather than wait days.

I guess it was a good call not merging this since it would have broken production I'm pretty sure.

**Comment by jonathanmetzman** [comment]
> > > @jonathanmetzman A gentle ping on this, as we need it to fix #8690 : ) Or would you prefer me to merge it into #8690 to test it out? Thanks!
> > 
> > 
> > I was offline on monday and tuesday. If there's an issue that doesn't effect production but is blocking you, feel free to merge it rather than wait days.
> 
> I guess it was a good call not merging this since it would have broken production I'm pretty sure.

No idea how trial build passed even though you had a comment that was not properly commented. Bash is such a disaster.

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [reply infra/base-images/base-runner/test_all.py:190]: Sorry I mean a capital letter obviously

Comment (Alan32Liu) [reply infra/base-images/base-runner/test_all.py:190]: Ah, I see. Thanks : )

**Comment by Alan32Liu** [comment]
> > > > @jonathanmetzman A gentle ping on this, as we need it to fix #8690 : ) Or would you prefer me to merge it into #8690 to test it out? Thanks!
> > > 
> > > 
> > > I was offline on monday and tuesday. If there's an issue that doesn't effect production but is blocking you, feel free to merge it rather than wait days.
> > 
> > 
> > I guess it was a good call not merging this since it would have broken production I'm pretty sure.
> 
> No idea how trial build passed even though you had a comment that was not properly commented. Bash is such a disaster.

Yeah, that is strange. I will address the comments above today and have it ready by your morning.
Thanks for the comments!

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

File: infra/base-images/base-runner/test_all.py
Comment (Alan32Liu) [reply infra/base-images/base-runner/test_all.py]: Done. I guess this is because we won't set `HELPER` if we are not building with `helper.py`, but is it safer to compare the value? (e.g. just in case some future commit sets it to something else.)

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [reply infra/base-images/base-runner/test_all.py]: Nevermind, just compare it to True so we'll handle "False" properly.

File: infra/base-images/base-runner/bad_build_check
Comment (jonathanmetzman) [inline infra/base-images/base-runner/bad_build_check:444]: What does `"${HELPER:-}"` do?

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [inline infra/base-images/base-runner/test_all.py:190]: Do = True again please.

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [inline infra/base-images/base-runner/test_all.py:198]: single quotes.

**Review by jonathanmetzman — APPROVED** [review]
lgtm

File: infra/base-images/base-runner/bad_build_check
Comment (Alan32Liu) [reply infra/base-images/base-runner/bad_build_check:444]: It [substitutes `HELPER` with an empty string](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) and avoids the '[unbounded variable error](https://bnikolic.co.uk/blog/bash-unbound-variable)' when `HELPER` is not set (e.g., when not using `helper.py`).
I choose this way because it is consistent with [other places](https://github.com/google/oss-fuzz/blob/86e6dc3e2ab6665711931f0c53bff8e0181d7114/infra/base-images/base-runner/bad_build_check#L312).

**Comment by jonathanmetzman** [comment]
sounds good!

File: infra/base-images/base-runner/test_all.py
Comment (Alan32Liu) [reply infra/base-images/base-runner/test_all.py:198]: I am not sure if I understand it correctly, but single quotes around `SANITIZER` will break the single quote of `f-sring`:
```python
  File "/usr/local/bin/test_all.py", line 198
    f'__centipede_{os.getenv('SANITIZER')}')
                              ^
SyntaxError: invalid syntax
```
Escaping single quotes is not allowed in `f-string` either:
```python
  File "/usr/local/bin/test_all.py", line 198
    f'__centipede_{os.getenv(\'SANITIZER\')}')
    ^
SyntaxError: f-string expression part cannot include a backslash
```

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine centipede

File: infra/base-images/base-runner/test_all.py
Comment (Alan32Liu) [reply infra/base-images/base-runner/test_all.py]: Ah, I just changed it back to `== True`.
Can we use 1 or 0, though? I thought we can only set env as strings.

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer none --fuzzing-engine centipede

File: infra/base-images/base-runner/test_all.py
Comment (jonathanmetzman) [reply infra/base-images/base-runner/test_all.py:198]: Oh nevermind. You were right initially and I was wrong.

**Comment by jonathanmetzman** [comment]
Let's wait until Monday your time to merge this so we can avoid a weekend breakage.
Have a good weekend!

**Comment by Alan32Liu** [comment]
/gcbrun trial_build.py all --sanitizer none --fuzzing-engine libfuzzer

**Pull request merged**
